mmdet/core/hook/ema.py

```diff
@@ -0,0 +1,121 @@
+    """
+
+    def __init__(self, skip_bn_running_stats=False, interval=1, resume_from=None):
+        self.skip_bn_running_stats = skip_bn_running_stats
+        self.interval = interval
+        self.checkpoint = resume_from
+
+    def before_run(self, runner):
+        """To resume model with it's ema parameters more friendly.
+        Register ema parameter as ``named_buffer`` to model
+        """
+        model = runner.model
+        if is_parallel(model):
+            model = model.module
+        self.param_ema_buffer = {}
+        if self.skip_bn_running_stats:
+            self.model_parameters = dict(model.named_parameters(recurse=True))
+        else:
+            self.model_parameters = model.state_dict()
+        for name, value in self.model_parameters.items():
+            # "." is not allowed in module's buffer name
+            buffer_name = f"ema_{name.replace('.', '_')}"
+            self.param_ema_buffer[name] = buffer_name
+            model.register_buffer(buffer_name, value.data.clone())
+        self.model_buffers = dict(model.named_buffers(recurse=True))
+        if self.checkpoint is not None:
+            runner.resume(self.checkpoint)
+
+    def get_momentum(self, runner):
+        raise NotImplementedError
+
+    def after_train_iter(self, runner):
+        """Update ema parameter every self.interval iterations."""
+        if runner.iter % self.interval != 0:
+            return
+        momentum = self.get_momentum(runner)
+        for name, parameter in self.model_parameters.items():
+            # exclude num_tracking
+            if parameter.dtype.is_floating_point:
+                buffer_name = self.param_ema_buffer[name]
+                buffer_parameter = self.model_buffers[buffer_name]
+                buffer_parameter.mul_(momentum).add_(parameter.data, alpha=1 - momentum)
+
+    def after_train_epoch(self, runner):
+        """We load parameter values from ema backup to model before the
+        EvalHook."""
+        self._swap_ema_parameters()
+
+    def before_train_epoch(self, runner):
+        """We recover model's parameter from ema backup after last epoch's
+        EvalHook."""
+        self._swap_ema_parameters()
+
+    def _swap_ema_parameters(self):
+        """Swap the parameter of model with parameter in ema_buffer."""
+        for name, value in self.model_parameters.items():
+            temp = value.data.clone()
+            ema_buffer = self.model_buffers[self.param_ema_buffer[name]]
+            value.data.copy_(ema_buffer.data)
+            ema_buffer.data.copy_(temp)
+
+
+@HOOKS.register_module()
+class ExpDecayEMAHook(BaseEMAHook):
+    """ Exponential decay EMAHook.
+    Args:
+        decay (float): Exponential decay coefficient. Default to 0.9998
+        total_iter (int): The total number of iterations of EMA decay.
+           Defaults to 100.
+    """
+    def __init__(self, decay=0.9998, total_iter=2000, **kwargs):
```

Comment:
Because we define `get_momentum` in the base class, Can we use a uniform argument `momentum` for all subclass instead of decay here

Reply:
use decay as tf and yolov5